www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,290 @@
+ 
+During failure recovery, we collect the state from all these sources and reconcile it into a consistent state using few simple rules. Lets look at how these rules play out in two scenarios - a failed database and a failed transaction coordinator. 
+
+### Recovering from database failure
+
+This is a case where the transaction coordinator lost contact with a database. The first order of business is to get back in touch. Unfortunately, Postgres does not have "database is back up" hook that we can use for a callback. Therefore both the coordinator and the `nile_ddl` extension periodically check the connection and attempt to reconnect if there is an issue. Once a connection is established, the transaction coordinator will check `pg_prepared_xact` and `distributed_transactions` tables on the recovering database. For every open transaction in the coordinator's in-memory state, the recovering DB can be in one of 3 states:
+
+- Transaction committed successfuly. In this case, the coordinator just needs to update its in-memory state and the metadata state table. 
+- Transaction prepared successfully but did not commit. In this case the coordinator checks the results of the transaction vote in its in-memory state, adding a "yes" vote from the recovered database. Based on the overall vote, the coordinator then decides if the transaction should commit or roll-back and distributes commands accordingly. 
+- Transaction does not appear in either of the database state tables. This means that this database already rolled back the transaction, and it never successfully prepared. This means that the database voted "no" in the commit vote, and the transaction coordinator has to distribute `ROLLBACK PREPARED` to the other databases that participated in the transaction.
+
+There is one special case for database failure, which is the case where the database where the transaction originated has failed. If you recall, this database is responsible for notifying the transaction coordinator that about the `XACT_EVENT_PRE_COMMIT` and `XACT_EVENT_COMMIT` events for the transaction. If the originating database fails, the coordinator may be stuck with an "orphan" transactions that never get prepared or committed. For this case, we rely on the transaction timeouts and a scheduled thread on the coordinator to roll back the orphaned transaction.
+
+### Recovering from coordinator failure
+
+If the coordinator itself fails, the first order of business is to reconstruct the in-memory state, so it will be aware of all the on-going transactions and capable of completing them. It first uses the latest state from the `dist_tx_participants`, and then start checking state of individual databases.
```

Comment:
We also use `distributed_transactions` on Metadata to retrieve all 'in-progress' transactions. We didn't mention this table above. We mentioned table with the similar name on normal PDBs, but that table is a source of truth only for transactions originating from that PDB and it knows nothing about transactions originating from other PDBs.